geti_sdk/deployment/deployed_model.py

```diff
@@ -623,6 +650,92 @@ def infer(self, image: np.ndarray, explain: bool = False) -> Prediction:
+            You can for instance pass a dictionary, or a tuple/list of objects.
+        """
+        if not self._async_callback_defined:
+            logging.warning(
+                "No callback function defined to handle asynchronous inference, "
+                "please make sure to define a callback using "
+                "`.set_asynchronous_callback`, otherwise your inference results may be "
+                "lost."
+            )
+        preprocessed_image, metadata = self._preprocess(image)
+        self._inference_model.infer_async_raw(
+            preprocessed_image,
+            {"explain": explain, "metadata": metadata, "runtime_data": runtime_data},
+        )
+
+    def set_asynchronous_callback(
+        self, callback_function: Callable[[Prediction, Optional[Any]], None]
+    ) -> None:
+        """
+        Set the callback function to handle asynchronous inference results. This
+        function is called whenever a result for an asynchronous inference request
+        comes available.
+
+        :param callback_function: Function that should be called to handle
+            asynchronous inference results. The function should take the following
+            input parameters:
+
+             1. The inference results (the Prediction). This is the primary input
+             2. Any additional data that will be passed with the infer
+                request at runtime. For example, this could be a timestamp for the
+                frame, or a title/filepath, etc. This can be in the form of any object:
+                You can for instance pass a dictionary, or a tuple/list of multiple
+                objects
+
+        """
+
+        def full_callback(infer_request, async_metadata: Dict[str, Any]):
+            # Basic postprocessing, convert to `Prediction` object
+            metadata = async_metadata["metadata"]
+            explain = async_metadata["explain"]
+            runtime_data = async_metadata["runtime_data"]
+
+            raw_result = self._inference_model.inference_adapter.get_raw_result(
+                infer_request
+            )
+            result = self._inference_model.postprocess(raw_result, metadata)
+
+            prediction = self._converter.convert_to_prediction(
+                result, image_shape=metadata["original_shape"]
+            )
+
+            # Add optional explainability outputs
+            if explain:
+                saliency_map, repr_vector = self._postprocess_explain_outputs(
+                    inference_results=result, metadata=metadata
+                )
+                prediction.feature_vector = repr_vector
+                result_medium = ResultMedium(name="saliency map", type="saliency map")
+                result_medium.data = saliency_map
+                prediction.maps = [result_medium]
+
+            # User defined callback to further process the prediction results
+            callback_function(prediction, runtime_data)
+
```

Comment:
this is a code duplication which adds support overhead, is there a way to use the same function in both cases?

Reply:
Agreed, updated